spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java

```diff
@@ -0,0 +1,262 @@
+                    "sort_order => 'id ASC none')", catalogName, tableIdent));
+
+    // Test for sort_order with invalid sort direction
+    AssertHelpers.assertThrows("Should reject calls with error message",
+            IllegalArgumentException.class, "Unable to parse sortOrder:",
+            () -> sql("CALL %s.system.rewrite_data_files(table => '%s', strategy => 'sort', " +
+                    "sort_order => 'id none NULLS FIRST')", catalogName, tableIdent));
+
+    // Test for sort_order with invalid column name
+    AssertHelpers.assertThrows("Should reject calls with error message",
+            ValidationException.class, "Cannot find field 'col1' in struct:" +
+                    " struct<1: id: required long, 2: data: optional string>",
+            () -> sql("CALL %s.system.rewrite_data_files(table => '%s', strategy => 'sort', " +
+                    "sort_order => 'col1 DESC NULLS FIRST')", catalogName, tableIdent));
+
+    // Test for sort_order with invalid filter column col1
+    AssertHelpers.assertThrows("Should reject calls with error message",
+            ValidationException.class, "Cannot find field 'col1' in struct:" +
+                    " struct<1: id: required long, 2: data: optional string>",
+            () -> sql("CALL %s.system.rewrite_data_files(table => '%s', " +
+                    "where => 'col1 = 3')", catalogName, tableIdent));
+  }
+
+  @Test
+  public void testInvalidCasesForRewriteDataFiles() {
+    AssertHelpers.assertThrows("Should not allow mixed args",
+        AnalysisException.class, "Named and positional arguments cannot be mixed",
+        () -> sql("CALL %s.system.rewrite_data_files('n', table => 't')", catalogName));
+
+    AssertHelpers.assertThrows("Should not resolve procedures in arbitrary namespaces",
+        NoSuchProcedureException.class, "not found",
+        () -> sql("CALL %s.custom.rewrite_data_files('n', 't')", catalogName));
+
+    AssertHelpers.assertThrows("Should reject calls without all required args",
+        AnalysisException.class, "Missing required parameters",
+        () -> sql("CALL %s.system.rewrite_data_files()", catalogName));
+
+    AssertHelpers.assertThrows("Should reject duplicate arg names name",
+        AnalysisException.class, "Duplicate procedure argument: table",
+        () -> sql("CALL %s.system.rewrite_data_files(table => 't', table => 't')", catalogName));
+
+    AssertHelpers.assertThrows("Should reject calls with empty table identifier",
+        IllegalArgumentException.class, "Cannot handle an empty identifier",
+        () -> sql("CALL %s.system.rewrite_data_files('')", catalogName));
+  }
+
+  private void createTable() {
+    sql("CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg", tableName);
+  }
+
+  private void createPartitionTable() {
+    sql("CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg PARTITIONED BY (data)", tableName);
+  }
+
+  private void insertData(int repeatCounter) {
```

Comment:
Repeat counter is not very clear, maybe this should just be numFiles or something like that?

Then the actual implementation should probably use a single dataframe write rather than an individual sql execution for each row.

Something like in the Action test suit

spark.range(0, numRows, 1, numFiles).write...

Reply:
yes, I will change to  `filesCount`. But I tried using dataframe, it interrupts with my logic of having specific number of files per partition. 
Can I keep the SQL ? SQL is easy for me.

Reply:
I think it is a lot less efficient here since not only do we write these files in serial, but we also have the cost of setting up a new spark for each file. It's best if we reduce the number of spark jobs we have to run in tests to keep test runtime down. So I would definitely consider rewriting it, as for the partitioning logic you can check the code in Action tests since we handle writing to multiple partitions there as well.

Reply:
more details on problem of using dataframe here https://github.com/apache/iceberg/pull/3375#discussion_r748691792

Reply:
I'm not sure what the links are doing but they do not go anywhere for me.

Reply:
ok, I was pointing to one of the old comments in the same pr, I have tagged you there now.

Reply:
My problem is not about writing multiple partition, but distributing files across partitions. so test case can expect fixed number of files in partition.

Reply:
The number of files in the partition will be equal to the number of Spark tasks, just like in the unpartitioned case where there is only a single partition as long as the number of records is greater than the number of partitions * the requested number of files per partition.